lib/dao.js

```diff
@@ -629,6 +629,176 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
 };
 
 /**
+ * Replace or insert a model instance: replace exiting record if one is found, such that parameter `data.id` matches `id` of model instance;
+ * otherwise, insert a new record.
+ *
+ * @param {Object} data The model instance data
+ * @param {Object} [options] Options for replaceOrCreate
+ * @param {Function} cb The callback function (optional).
+ */
+
+DataAccessObject.replaceOrCreate = function replaceOrCreate(data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) {
+    return connectionPromise;
+  }
+  assert((options || cb), 'The cb must be passed');
+
+  if (cb === undefined) {
+    if (typeof options === 'function') {
+      cb = options;
+      options = {};
+    }
+  }
+
+  cb = cb || utils.createPromiseCallback();
+  data = data || {};
+  options = options || {};
+
+  assert(typeof data === 'object', 'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var hookState = {};
+
+  var self = this;
+  var Model = this;
+  var connector = Model.getConnector();
+
+  var id = getIdValue(this, data);
+  if (id === undefined || id === null) {
+    return this.create(data, options, cb);
+  }
+  
+  var inst;
+  if (data instanceof Model) {
+    inst = data;
+  } else {
+    inst = new Model(data);
+  }
+  
+  var context = {
+    Model: Model,
+    query: byIdQuery(Model, id),
+    hookState: hookState,
+    options: options
+  };
+  Model.notifyObserversOf('access', context, doReplaceOrCreate);
+
+  function doReplaceOrCreate(err, ctx) {
+    if (err) return cb(err);
+
+    var isOriginalQuery = isWhereByGivenId(Model, ctx.query.where, id);
+    var where = ctx.query.where;
+    if (connector.replaceOrCreate && isOriginalQuery) {
+      var context = {
+        Model: Model,
+        instance: inst,
+        hookState: hookState,
+        options: options
+      };
+      Model.notifyObserversOf('before save', context, function(err, ctx) {
+        if (err) return cb(err);
+
+        var update = inst.toObject(false);
+        Model.applyProperties(update, inst);
+        Model = Model.lookupModel(update);
+
+        var connector = self.getConnector();
+
+        // only when options.validate is not set, take model-setting into consideration
+        if (options.validate === undefined && Model.settings.automaticValidation === false) {
+          return callConnector();
+        }
+
+        inst.isValid(function(valid) {
+          if (!valid) {
+              return cb(new ValidationError(inst), inst);
+          }
+          callConnector();
+        }, update);
+
+        function callConnector() {
+          update = removeUndefined(update);
+          context = {
+            Model: Model,
+            where: where,
+            data: update,
+            currentInstance: inst,
```

Comment:
As commented in tests, you should provide `instance: inst` instead of `where+data+currentInstance`.